### ISSUE TITLE
Minecart gets too much momentum

### ISSUE THREAD

**Issue by henrikx** [ergor/hsrails#6]
It seems that after passing a high speed rail section, the minecart takes a lot of time to slow down. It gets way too much momentum.
Minecraft version 1.19 PaperMC build 8.

**Comment by henrikx** [comment]
Using a normal powered rail to slow down for a turn, and then putting a high speed rail right after it also gives the minecart an unatural amount of acceleration.

**Comment by ergor** [comment]
What speed multiplier are you using?
You may need more than 1 normal powered rail to slow down sufficiently

Regarding the second comment, I'm not sure what is meant by unnatural amount of acceleration. Can you elaborate what happens?

**Comment by henrikx** [comment]
> What speed multiplier are you using? You may need more than 1 normal powered rail to slow down sufficiently
> 
> Regarding the second comment, I'm not sure what is meant by unnatural amount of acceleration. Can you elaborate what happens?

This is with the default multiplier of 4.

I put one or two normal powered-rails before a turn to slow down. Whereas normally there needs to be 3-4 high-speed powered rails to get to max speed, there only has to be 1 after the turn to get back to the previous top speed.

**Comment by henrikx** [comment]
That explanation makes sense. Maybe it would be possible to have the plugin reduce it's momentum when passing over a normal powered rail?

Anyways, it's mostly just an oddity.

**Comment by ergor** [comment]
I took a quick look in the Spigot API, but there doesn't seem to be a direct way to reduce momentum. There is the setVelocity method, maybe that could be used. I'll leave the issue open, maybe I can fix it some time. Otherwise it could just be noted in the description of the plugin.

Thanks for letting me know!

**Comment by ergor** [comment]
Hi @henrikx!

I recently released version 1.3.0 that added hardBrakeBlocks. It leverages the setVeloctiy method I mentioned above. If you try using that, I think the issue should be solved

**Comment by Janicki92** [comment]
I am also running into this issue.  I have discovered that if you have too many powered rails (i.e. a track that starts off going up a large incline with regular powered rails) and then by hitting just 1 High-Speed Powered Rail, it will immediately start going max speed of 32m/s and it will **never** slow down.  We made a straight-line track that went over 3000 blocks with the only powered & high-speed powered rails at the very beginning of the track to get going.  Then the cart **never** loses it's speed/acceleration.  We could have made that track go forever and the cart would've maintained 32m/s.  

Is there any way to fix this without just adding a brake to the track?  For an SMP, we're looking to have faster minecarts but still within reason and should cost more materials the longer the track / the faster you want it to go.  Also it becomes difficult sometimes dealing with this "bug" when you're not trying to have the cart go full speed right off the bat.

**Comment by ergor** [comment]
Thanks @Janicki92 , I will try to replicate this and see if something can be done about it

**Comment by ergor** [comment]
I have found the root cause of the bug after some digging.

What happens is: when the cart has its max speed set higher than vanilla, it enters a weird, infinite cycle where it rapidly oscillates between two velocities. This means it never slows down, because the velocity jumps up and down and back up.

I dived into the decompiled Minecraft code that handles Minecart movement. It is a hot mess full of edge cases. Undoubtedly, some of these edge cases introduces this oscillation.

Minecart momentum works as follows (vanilla values):
- Max speed is 0.4 meters per tick.
- The cart has some velocity, which actually can _exceed_ the max speed.
- This is the "momentum" the cart has. The maximum velocity is 2.054, but the max speed is 0.4.
- What the game does is, it is always decreasing the velocity, while clamping the actual movement to at most 0.4 meters each tick.

In other words, the cart will move 0.4 meters each tick as long as its velocity is higher than or equal to 0.4. The velocity is always decreasing, but the "rate of movement" stays constant at velocities above 0.4. When the velocity drops below 0.4, the cart starts visibly decelerating.

When messing with the max speed (as this plugin does), something breaks in the Minecraft code (again, probably due to the plethora of edge case handling in the Minecraft code).

What I had to do was keep track of the cart state separately to perform my own momentum calculations, and progressively decrease max speed when I detect that the cart is coasting. 

I will publish the jar shortly.

Feedback and testing would be greatly appreciated!

**Comment by Janicki92** [comment]
Awesome!  I'm not sure how to compile this myself but once the .jar file has been posted I would be happy to have my server start testing right away

**Comment by ergor** [comment]
Great! I uploaded it to bukkit on wednesday, but it is still under review. I don't know when it will be completed.

So I have uploaded it to my website in the meantime if interested: [download link](https://www.netb.no/tmp/hsrails.jar)
